t3f/autodiff.py

```diff
@@ -0,0 +1,304 @@
+  if left is None:
+    left = decompositions.orthogonalize_tt_cores(tt)
+  if right is None:
+    right = decompositions.orthogonalize_tt_cores(left, left_to_right=False)
+  left_tangent_tt_ranks = shapes.lazy_tt_ranks(left)
+  right_tangent_tt_ranks = shapes.lazy_tt_ranks(left)
+  raw_shape = shapes.lazy_raw_shape(left)
+  right_rank_dim = left.right_tt_rank_dim
+  left_rank_dim = left.left_tt_rank_dim
+  is_batch_case = hasattr(tt, 'batch_size')
+  for i in range(num_dims):
+    left_tt_core = left.tt_cores[i]
+    right_tt_core = right.tt_cores[i]
+
+    if i == 0:
+      tangent_core = tf.concat((deltas[i], left_tt_core),
+                               axis=right_rank_dim)
+    elif i == num_dims - 1:
+      tangent_core = tf.concat((right_tt_core, deltas[i]),
+                               axis=left_rank_dim)
+    else:
+      rank_1 = right_tangent_tt_ranks[i]
+      rank_2 = left_tangent_tt_ranks[i + 1]
+      if tt.is_tt_matrix():
+        mode_size_n = raw_shape[0][i]
+        mode_size_m = raw_shape[1][i]
+        shape = [rank_1, mode_size_n, mode_size_m, rank_2]
+      else:
+        mode_size_n = raw_shape[0][i]
+        shape = [rank_1, mode_size_n, rank_2]
+      if is_batch_case:
+        shape = [tt.batch_size] + shape
+      zeros = tf.zeros(shape, dtype=dtype)
+      upper = tf.concat((right_tt_core, zeros), axis=right_rank_dim)
+      lower = tf.concat((deltas[i], left_tt_core), axis=right_rank_dim)
+      tangent_core = tf.concat((upper, lower), axis=left_rank_dim)
+    cores.append(tangent_core)
+  # Create instance of the same class as tt.
+  tangent = tt.__class__(cores)
+  tangent.projection_on = tt
+  return tangent
+
+
+def _gradients(func, x, x_projection, left, right):
+  """Internal version of t3f.gradients that assumes some precomputed inputs."""
+  h = func(x_projection)
+  cores_grad = tf.gradients(h, x_projection.tt_cores)
+  deltas = []
+  for i in range(x.ndims()):
+    if x.is_tt_matrix():
+      r1, n, m, r2 = left.tt_cores[i].shape.as_list()
+    else:
+      r1, n, r2 = left.tt_cores[i].shape.as_list()
+    q = tf.reshape(left.tt_cores[i], (-1, r2))
+    if x.is_tt_matrix():
+      if i == 0:
+        curr_grad = cores_grad[i][:, :, :, :r2]
+      elif i == x.ndims() - 1:
+        curr_grad = cores_grad[i][r1:, :, :, :]
+      else:
+        curr_grad = cores_grad[i][r1:, :, :, :r2]
+    else:
+      if i == 0:
+        curr_grad = cores_grad[i][:, :, :r2]
+      elif i == w.ndims() - 1:
+        curr_grad = cores_grad[i][r1:, :, :]
+      else:
+        curr_grad = cores_grad[i][r1:, :, :r2]
+    if i < x.ndims() - 1:
+      delta = curr_grad
+      delta = tf.reshape(delta, (-1, r2))
+      delta -= tf.matmul(q, tf.matmul(tf.transpose(q), delta))
+      delta = tf.reshape(delta, left.tt_cores[i].shape)
+    else:
+      delta = curr_grad
+    deltas.append(delta)
+  return deltas_to_tangent_space(deltas, x, left, right)
+
+
+def gradients(func, x, name='t3f_gradients'):
```

Comment:
Add somehting like debug=True which checks that f(left_orthogonalize_x) ~= f(right_orthogonalize_x)